pkg/testcase/clusterrestore.go

```diff
@@ -0,0 +1,260 @@
+	for _, listObject := range s3List {
+		if strings.Contains(*listObject.Key, onDemandPath) {
+			shared.LogLevel("info", "snapshot found: %s", onDemandPath)
+			break
+		}
+	}
+
+	shared.LogLevel("info", "successfully validated s3 snapshot save in s3")
+}
+
+func stopInstances(cluster *shared.Cluster, ec2 *aws.Client) {
+	var instancesIPs []string
+
+	instancesIPs = append(instancesIPs, cluster.ServerIPs...)
+	instancesIPs = append(instancesIPs, cluster.AgentIPs...)
+
+	for _, ip := range instancesIPs {
+		id, idsErr := ec2.GetInstanceIDByIP(ip)
+		Expect(idsErr).NotTo(HaveOccurred())
+		err := ec2.StopInstance(id)
+		Expect(err).NotTo(HaveOccurred())
+	}
+}
+
+func installProduct(cluster *shared.Cluster, newClusterIP, version string) {
+	setConfigFile(cluster, newClusterIP)
+
+	installCmd := shared.GetInstallCmd(cluster.Config.Product, version, "server")
+	if cluster.Config.Product == "k3s" {
+		skipInstall := fmt.Sprintf(" INSTALL_%s_SKIP_ENABLE=true ", strings.ToUpper(cluster.Config.Product))
+		installCmd = strings.Replace(installCmd, "sh", skipInstall+" "+"  sh", 1)
+	}
+
+	_, installCmdErr := shared.RunCommandOnNode(installCmd, newClusterIP)
+	Expect(installCmdErr).NotTo(HaveOccurred())
+
+	shared.LogLevel("info", "%s successfully installed on server: %s", cluster.Config.Product, newClusterIP)
+}
+
+func setConfigFile(cluster *shared.Cluster, newClusterIP string) {
+	serverFlags := os.Getenv("server_flags")
+	if serverFlags == "" {
+		serverFlags = "write-kubeconfig-mode: 644"
+	}
+	serverFlags = strings.ReplaceAll(serverFlags, `\n`, "\n")
+
+	tempFilePath := "/tmp/config.yaml"
+	tempFile, err := os.Create(tempFilePath)
+	Expect(err).NotTo(HaveOccurred())
+
+	defer tempFile.Close()
+
+	_, writeErr := fmt.Fprintf(tempFile, "node-external-ip: %s\n", newClusterIP)
+	Expect(writeErr).NotTo(HaveOccurred())
+
+	flagValues := strings.Split(serverFlags, "\n")
+	for _, entry := range flagValues {
+		entry = strings.TrimSpace(entry)
+		if entry != "" {
+			_, err := fmt.Fprintf(tempFile, "%s\n", entry)
+			Expect(err).NotTo(HaveOccurred())
+		}
+	}
+
+	remoteDir := fmt.Sprintf("/etc/rancher/%s/", cluster.Config.Product)
+	user := os.Getenv("aws_user")
+	cmd := fmt.Sprintf("sudo mkdir -p %s && sudo chown %s %s ", remoteDir, user, remoteDir)
+
+	_, mkdirCmdErr := shared.RunCommandOnNode(cmd, newClusterIP)
+	Expect(mkdirCmdErr).NotTo(HaveOccurred())
+
+	scpErr := shared.RunScp(cluster, newClusterIP, []string{tempFile.Name()}, []string{remoteDir + "config.yaml"})
+	Expect(scpErr).NotTo(HaveOccurred())
+}
+
+func restoreS3Snapshot(
+	cluster *shared.Cluster,
+	onDemandPath,
+	token string,
+	newClusterIP string,
+	flags *customflag.FlagConfig,
+) {
+	var (
+		restoreCmdRes string
+		restoreCmdErr error
+	)
+
+	productLocationCmd, findErr := shared.FindPath(cluster.Config.Product, newClusterIP)
+	Expect(findErr).NotTo(HaveOccurred())
+
+	restoreCmd := fmt.Sprintf("sudo %s server --cluster-reset --etcd-s3 --cluster-reset-restore-path=%s"+
```

Comment:
Should we add in the config.yaml of the new instance? That way we dont have to run this long command

Reply:
hum yeah but i like this way looks more the " way " i use to do manually

Reply:
Right it is one of the way, but we should also add test to do it via config.yaml, may be a improvement

Reply:
agree